### PR TITLE
aligned servicePath: none behaviour with the one of orion

### PR DESCRIPTION
**Pull request by pooja1pathak**
## Proposed changes

aligned servicePath: none behaviour with the one of orion. fix for issue #563

## Types of changes

What types of changes does your code introduce to the project: _Put
an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
      functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after
creating the PR. If you're unsure about any of them, don't hesitate to
ask. We're here to help! This is simply a reminder of what we are going
to look for before merging your code._

- [x] I have read the [CONTRIBUTING][contrib] doc
- [x] I have signed the [CLA][cla]
- [ ] I have added tests that prove my fix is effective or that my
      feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in
      downstream modules

## Further comments

If this is a relatively large or complex change, kick off the discussion
by explaining why you chose the solution you did and what alternatives
you considered, etc...

[cla]: https://raw.githubusercontent.com/orchestracities/ngsi-timeseries-api/master/individual-cla.pdf
    "Martel Open Source Software Individual Contributor License Agreement"
[contrib]: https://github.com/smartsdk/ngsi-timeseries-api/blob/master/CONTRIBUTING.md
    "Contributing to QuantumLeap"

### CODE REVIEW

**Comment by pooja1pathak** [comment]
@chicco785 in ```orion``` ```Only /absolute/ Service Paths allowed [a service path must begin with /]```. Do we need to implement exact behavior in QL also? If yes, then I will need to update test cases and add a check for ```fiware-servicePath```.

For failing translator test case, if query is for specific fiware-servicePath then only the entities related to that fiware-servicePath should be returned. So, in these cases no entity was added in ```fiware_servicepath="/eu"``` so 0 should be the outcome.

Reporter tests failing reason is still unknown to me. As majority of test cases are failing for ```service: t2```. Please guide me if you have any points to resolve this issue.

File: src/translators/sql_translator.py [hunk 1]
```diff
@@ -1551,10 +1551,10 @@ def query_entity_types(self, fiware_service=None, fiware_servicepath=None):
             for et in all_types:
                 stmt = "select distinct(entity_type) from {}".format(et)
                 if fiware_servicepath == '/':
-                    stmt = stmt + " WHERE {} ~* '/.*'" \
+                    stmt = stmt + " WHERE {} = '/'" \
```
Comment (chicco785) [inline src/translators/sql_translator.py:1554]: not sure about this changes

File: src/translators/sql_translator.py [hunk 2]
```diff
@@ -1551,10 +1551,10 @@ def query_entity_types(self, fiware_service=None, fiware_servicepath=None):
...
                         .format(FIWARE_SERVICEPATH)
                 elif fiware_servicepath:
-                    stmt = stmt + " WHERE {} ~* '{}($|/.*)'" \
+                    stmt = stmt + " WHERE {} = '{}'" \
```
Comment (chicco785) [inline src/translators/sql_translator.py:1557]: not sure about this changes

Comment (chicco785) [reply src/translators/sql_translator.py:1557]: most probably if your revert this, tests should pass

File: src/translators/sql_translator.py (hunk 1 above)
Comment (chicco785) [reply src/translators/sql_translator.py:1554]: most probably if your revert this, tests should pass

Comment (pooja1pathak) [reply src/translators/sql_translator.py]: I have removed ```.*``` as this results in extracting all the entities. I have also updated the if block as the statement can be used in both the cases.

File: src/translators/sql_translator.py (hunk 2 above)
Comment (pooja1pathak) [reply src/translators/sql_translator.py]: I have removed ```.*``` as this results in extracting all the entities. I have also updated the if block as the statement can be used in both the cases.

File: src/translators/sql_translator.py (hunk 1 above)
Comment (pooja1pathak) [reply src/translators/sql_translator.py]: @chicco785 ```.*``` should not be the behavior expected from QL if we want it to behave in the same manner as that of ```orion```. Please consider my comment  https://github.com/orchestracities/ngsi-timeseries-api/pull/564#issuecomment-920820646 and let me know if I need to updated test cases in translator to match this behavior.

File: src/translators/sql_translator.py [hunk 3]
```diff
@@ -1550,11 +1550,8 @@ def query_entity_types(self, fiware_service=None, fiware_servicepath=None):
 
             for et in all_types:
                 stmt = "select distinct(entity_type) from {}".format(et)
-                if fiware_servicepath == '/':
-                    stmt = stmt + " WHERE {} ~* '/.*'" \
-                        .format(FIWARE_SERVICEPATH)
-                elif fiware_servicepath:
-                    stmt = stmt + " WHERE {} ~* '{}($|/.*)'" \
+                if fiware_servicepath:
+                    stmt = stmt + " WHERE {} ~* '{}'" \
```
Comment (chicco785) [inline src/translators/sql_translator.py:1554]: i am not sure about the change, failing tests are now more than before...

File: src/translators/sql_translator.py [hunk 4]
```diff
@@ -821,11 +821,11 @@ def _get_where_clause(self, entity_ids, from_date, to_date, fiware_sp=None,
             # Match prefix of fiware service path
             if fiware_sp == '/':
                 clauses.append(
-                    " " + prefix + FIWARE_SERVICEPATH + " ~* '/.*'")
+                    " " + prefix + FIWARE_SERVICEPATH + " = '/'")
```
Comment (chicco785) [inline src/translators/sql_translator.py:824]: and most probably also this query and the following one need to be fixed

File: src/translators/sql_translator.py (hunk 1 above)
Comment (pooja1pathak) [reply src/translators/sql_translator.py]: I have reverted my changes. We can consider points which I have mentioned in https://github.com/orchestracities/ngsi-timeseries-api/pull/564#issuecomment-920820646 in separate issue so its functionality matches ```orion``` broker. Or you have any suggestion please correct me.

**Comment by chicco785** [comment]
@c0c0n3 made a major refactor of the tests to wait for data consolidation in crate. can you rebase your pr?

**Comment by pooja1pathak** [comment]
@chicco785 @c0c0n3 as the tests are failing for timescale db, I need to do some debugging with timescale. 
Can you please guide me how can I set default database to timescale to run QL with timescale only.

**Comment by c0c0n3** [comment]
hi @pooja1pathak, here's an example of a docker compose file w/ QL+Timescale as a backend

```yaml
version: '3'

services:

  redis:
    image: redis:6.2.3
    ports:
      - "6379:6379"
    networks:
      - e2etest

  timescale:
    image: timescale/timescaledb-postgis:1.7.5-pg12
    ports:
      - "5432:5432"
    networks:
      - e2etest
    environment:
      - POSTGRES_PASSWORD=*

  quantumleap-db-setup:
    image: orchestracities/quantumleap-pg-init
    depends_on:
      - timescale
    networks:
      - e2etest
    environment:
      - QL_DB_PASS=*
      - QL_DB_INIT_DIR=/ql-db-init
      - PG_HOST=timescale
      - PG_PASS=*

  quantumleap:
    image: orchestracities/quantumleap
    depends_on:
      # - crate
      - timescale
      - redis
    networks:
      - e2etest
    ports:
      - "8668:8668"
    environment:
      - QL_DEFAULT_DB=timescale
      - POSTGRES_HOST=timescale
      - POSTGRES_PORT=5432
      - POSTGRES_DB_NAME=quantumleap
      - POSTGRES_DB_USER=quantumleap
      - POSTGRES_DB_PASS=*
      - POSTGRES_USE_SSL=False
      - CRATE_HOST=crate
      - USE_GEOCODING=False
      - CACHE_QUERIES=True
      - REDIS_HOST=redis
      - REDIS_PORT=6379
      - LOGLEVEL=DEBUG

networks:
  e2etest:
    driver: bridge
```

I haven't tested it, so you might have to tweak something---I used my famed C&P tech to whip it together :-)

**Comment by pooja1pathak** [comment]
@c0c0n3 Thank you for your response. I need to add timescale db with the source setup, is there any config file which I can edit for QL_DEFAULT_DB.

**Comment by pooja1pathak** [comment]
@c0c0n3 @chicco785 I have checked tables created in timescale db with QL notify API, it inserts "/" when no fiware-servicePath is mentioned. I am also able to retrieve data using QL APIs. It is working same as that in case of crate DB but still test cases are failing.

If you can spare sometime and look into the reason behind failing test cases and guide me to update this PR.

**Comment by c0c0n3** [comment]
Hi @pooja1pathak, thank you so much for all the effort you put into this PR so far, as always we do appreciate that. I had a quick look at some of the tests that fail, e.g.

- https://app.circleci.com/pipelines/github/pooja1pathak/ngsi-timeseries-api/2/workflows/efcad659-730b-47ab-b9c8-eccacc53c6ce/jobs/25

and it could be that there's some dirty state from previous tests that make tests executed later fail. (We know writing tests that way is not a good idea to start with, but for various reasons we ended up with alot of hidden global state deps among what were initially supposed to be unit tests.) Can you try running each failing test in isolation and see if it still fails? For example, can you try this

```console
$ source setup_dev_env.sh
$ cd src/translators/tests/
$ export QL_CONFIG='ql-config.yml'
$ pytest test_delete.py -k 'test_delete_entities_with_tenancy'
$ unset QL_CONFIG
```

Does it still fail?

**Comment by pooja1pathak** [comment]
@c0c0n3  I have followed the method as you have suggested and tested it for ```test_delete.py```, test case is passed
image

However I am unable to test reporter test cases using the same method. Do I need to create docker containers for these test case.

How can I clean the state to make it successful on CircleCI.

**Comment by c0c0n3** [comment]
@pooja1pathak first off, thanks alot for all the effort you're putting into this, much appreciated.

>  test_delete.py, test case is passed

Oh dear. I hoped that'd fail. So we've likely got some global state laying around that makes some tests fail. Can you run all the translator tests and see if you can reproduce the Circle CI failure? Here's how to run those tests on your box

```console
$ source setup_dev_env.sh
$ cd src/translators/tests
$ sh run_tests.sh
```

> to test reporter test cases

yah, for that, you can do this. Open a new terminal in the repo root, then

```console
$ source setup_dev_env.sh
$ cd src/reporter/tests
$ sh run_tests.sh
```

Hopefully, they should fail in a way similar to that we saw on CircleCI.

> How can I clean the state to make it successful on CircleCI.

Every time we trigger a build, CircleCI actually starts from a clean slate. I think the problem is, like I said, most of our tests unfortunately make use of global state (e.g. data in Crate) so there could be some dirty state from previous tests that make tests executed later fail---e.g. one tests writes an entity to a table `Room`, that entity doesn't get deleted, the next test writes another entity to the same table, then counts the rows expecting a count of 1 being returned whereas the actual count is 2.

**Comment by pooja1pathak** [comment]
@c0c0n3 I have run the tests (translator and reporter) in my local environment, the same test cases are failing in my local environment as in CircleCI.

> Every time we trigger a build, CircleCI actually starts from a clean slate. I think the problem is, like I said, most of our tests unfortunately make use of global state (e.g. data in Crate) so there could be some dirty state from previous tests that make tests executed later fail---e.g. one tests writes an entity to a table Room, that entity doesn't get deleted, the next test writes another entity to the same table, then counts the rows expecting a count of 1 being returned whereas the actual count is 2.

Yes I understand your concern. This may be the issue with translator tests (specifically with timescale).
However for reporter tests, on investigating I found that data for 'fiware-service: t2' is not inserted in the db, Although the response of POST request is ok. I am still looking into the issue.

**Comment by c0c0n3** [comment]
hi @pooja1pathak 

> I found that data for 'fiware-service: t2' is not inserted in the db.

Is that data getting inserted into Crate? How do you run the reporter tests? If you run them manually, you've got to make sure QL [loads][load] the [multitenant-backend config][cfg]. That could be one possible reason why t2's data doesn't get written to Timescale, but I'd need to know what your test env looks like exactly to be sure.



[cfg]: https://github.com/orchestracities/ngsi-timeseries-api/blob/master/src/reporter/tests/ql-config.yml
[load]: https://github.com/orchestracities/ngsi-timeseries-api/blob/master/src/reporter/tests/run_tests.sh#L11

**Comment by pooja1pathak** [comment]
@c0c0n3 I have updated the PR, now tests are passing. Can you please review it and let me know in case any other change is required from my side.

**Comment by chicco785** [comment]
@pooja1pathak thanks a lot, i will look into this today!

File: RELEASE_NOTES.md [hunk 5]
```diff
@@ -39,6 +39,7 @@
 - Added instanceId for each entry received (#565)
 - Support CrateDB authentication (#474)
 - Updated PG8000 to 1.23.0 (#586)
+- aligned servicePath: none behaviour with the one of orion (#564)
```
Comment (chicco785) [inline RELEASE_NOTES.md:42]: aligned servicePath: none behaviour with the one of orion (#564)
->
Aligned missing `Fiware-servicePath` behaviour with the one of Orion Context Broker (#564). This is a breaking change! Before no value for `Fiware-servicePath` was interpreted as python `None`, from now on, `None` will be replaced with `/`. This affects only users that manually injected data, since Orion, assume `/` when no servicePath is passed.

File: src/reporter/delete.py [hunk 6]
```diff
@@ -34,15 +34,15 @@ def delete_entity(entity_id, type_=None, from_date=None, to_date=None):
...
                              fiware_service=fiware_s())
             logging.getLogger(__name__).info(
                 "dropped entity_type {}".format(entity_type))
             return 'entity table dropped', 204
         elif drop_table:
             e = fiware_sp().split(",")
-            if e is not None and len(e) > 0:
+            if e != '/' and len(e) > 0:
                 return "request has one or more fiware service paths," \
                        "dropTable requires no service path", 422
```
Comment (chicco785) [inline src/reporter/delete.py:47]:
```
return "request has one or more fiware service paths," \ 
            "dropTable requires no service path", 422
```

should be replaced with

```
return "dropTable requires the root fiware-servicepath: '/'", 422
```

**Review by chicco785 — CHANGES_REQUESTED** [review]
Minor changes, and then for me is good to go. @c0c0n3 any additional item?

File: src/reporter/delete.py [hunk 7]
```diff
@@ -34,15 +34,15 @@ def delete_entity(entity_id, type_=None, from_date=None, to_date=None):
...
+        if drop_table and fiware_sp() == '/':
             trans.drop_table(etype=entity_type,
                              fiware_service=fiware_s())
             logging.getLogger(__name__).info(
                 "dropped entity_type {}".format(entity_type))
             return 'entity table dropped', 204
         elif drop_table:
             e = fiware_sp().split(",")
-            if e is not None and len(e) > 0:
+            if e != '/' and len(e) > 0:
```
Comment (c0c0n3) [inline src/reporter/delete.py:45]: I think there's a little problem with this line. `e` is a list not a string?

```python
>>> e = "/".split(",")
>>> e
['/']
>>> e != '/'
True
```

It looks like the logic should rather be (pseudo code)

    if len(e) > 1
       error 'too many svc paths'
    elif len == 0 or e[1] != '/'
       error 'can only drop if svc path = /'
    else
       drop table

@chicco785 is that right?
@pooja1pathak if @chicco785 confirms, then can we add some tests to check the branching logic?
thanks guys!!

Comment (c0c0n3) [reply src/reporter/delete.py:45]: yep, not sure either why we had those extra checks to start with. to me too, it'd be much better to just do what you're suggesting

```
if drop and fiware_sp() == '/':
   drop table
else:
   error ....
```

@pooja1pathak thoughts?

Comment (chicco785) [reply src/reporter/delete.py:45]: if drop_table and fiware_sp() == '/':
   drop!
elif drop_table:
    e = fiware_sp().split(",")
    if  e.count(e[0]) == len(e) and e[0] == '/':
       drop!
    else:
        return "dropTable requires the root fiware-servicepath: '/'", 422

Comment (pooja1pathak) [reply src/reporter/delete.py:45]: @c0c0n3 I agree with @chicco785 's suggestion. I will update the code as per above logic.

Comment (c0c0n3) [reply src/reporter/delete.py:45]: cool, bananas! Just one thing though. Isn't the above equivalent to this? that is @chicco785's initial suggestion:

```
if drop_table and fiware_sp() == '/':
   drop_logic
elif drop_table:
        return "dropTable requires the root fiware-servicepath: '/'", 422
```

I mean if you get into the second branch of the if statement, then you know `dropTable == True`, so it must be the case that `fiware_sp() != '/'` otherwise the first if statement would've been true? Am I missing something?

Comment (c0c0n3) [reply src/reporter/delete.py:45]: @chicco785 @pooja1pathak sorry ignore my comment above. I wrote that too quickly. I think @chicco785 wants to drop the table also in the case there's multiple service paths but the first one is `/`. E.g. if we get

```
/v2/...?dropTable=true
fiware-servicepath: /,/other
```

then we should still drop the table. @chicco785 is that right? or have I misunderstood the intention?

Comment (chicco785) [reply src/reporter/delete.py:45]: may idea was if we get `/, /other` we SHOULD NOT drop the table, if we get `/, /, /` we SHOULD drop the table

Comment (chicco785) [reply src/reporter/delete.py:45]: it got inspired for a quick solution to check that all elements in the split array are equal from this discussion:
https://stackoverflow.com/questions/3844801/check-if-all-elements-in-a-list-are-identical

But it may be wrong :)

Comment (pooja1pathak) [reply src/reporter/delete.py]: I have updated the PR as suggested

File: RELEASE_NOTES.md (hunk 5 above)
Comment (pooja1pathak) [reply RELEASE_NOTES.md]: updated

**Comment by c0c0n3** [comment]
@pooja1pathak thank you so much for all the effort you put into this PR, great job! Just one last snag: the markdown linter in circle CI is moaning about some formatting in the release notes

> RELEASE_NOTES.md:12:[7](https://github.com/orchestracities/ngsi-timeseries-api/runs/5329797352?check_suite_focus=true#step:4:7)1 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
> RELEASE_NOTES.md:49 MD012/no-multiple-blanks Multiple consecutive blank lines [Expected: 1; Actual: 2]

to reproduce on your box, run the `lint.md.sh` script. If you could please fix the markdown formatting so we can merge this PR. Thanks a million!

**Review by chicco785 — CHANGES_REQUESTED** [review]
As mentioned by @c0c0n3, current code tests only what happens when service_path = '/',
we should also test what happens when service path is None that the behavior implies servicePath = '/'.

**Comment by pooja1pathak** [comment]
> to reproduce on your box, run the `lint.md.sh` script. If you could please fix the markdown formatting so we can merge this PR. Thanks a million!

Done

File: RELEASE_NOTES.md [hunk 8]
```diff
@@ -41,6 +46,7 @@
 - Support CrateDB authentication (#474)
 - Updated PG8000 to 1.23.0 (#586)
 
+
```
[the remark this answers is not in the document]
Comment (pooja1pathak) [reply RELEASE_NOTES.md]: removed

File: src/reporter/httputil.py [hunk 9]
```diff
@@ -12,11 +12,18 @@ def fiware_sp() -> str:
...
 
 def fiware_correlator() -> str:
     """
     :return: The content of the FIWARE correlator path header if any.
     """
     return request.headers.get('Fiware-Correlator', None)
+
+
+def is_root_service_path() -> bool:
+    sp = fiware_sp() or ''  # cater for header not present
```
[the remark this answers is not in the document]
Comment (pooja1pathak) [reply src/reporter/httputil.py:26]: fiware_sp() will return '/', so I don't think this will create any issue.

[the remark this answers is not in the document]
Comment (chicco785) [reply src/reporter/httputil.py:26]: then we can remove `or '/'` no?

[the remark this answers is not in the document]
Comment (c0c0n3) [reply src/reporter/httputil.py:26]: @pooja1pathak @chicco785 yes you're both right, my mistake sorry :-)
this line

```
sp = fiware_sp() or ''
```

should rather be

```
sp = fiware_sp()
```

[the remark this answers is not in the document]
Comment (pooja1pathak) [reply src/reporter/httputil.py]: removed ''

**Comment by c0c0n3** [comment]
@pooja1pathak there's something fishy w/ git commits...it looks like these two files will be added if we merge this PR

- lint.md.sh
- lint.py.sh

and they will override the ones in `master` with empty files? Is there any way to remove them from this PR?

Also let us know what you'd like to do about adding more tests as @chicco785 [suggested][suggestion]. Do you still have time to spend on this or should it rather be something we an another PR?

Thanks!

[suggestion]: https://github.com/orchestracities/ngsi-timeseries-api/pull/564#pullrequestreview-893362658

**Comment by pooja1pathak** [comment]
I've updated the PR.

> Also let us know what you'd like to do about adding more tests as @chicco785 [suggested](https://github.com/orchestracities/ngsi-timeseries-api/pull/564#pullrequestreview-893362658). Do you still have time to spend on this or should it rather be something we do in an another PR?

I can work as @chicco785 suggested in another PR.

**Comment by chicco785** [comment]
> I've updated the PR.
> 
> > Also let us know what you'd like to do about adding more tests as @chicco785 [suggested](https://github.com/orchestracities/ngsi-timeseries-api/pull/564#pullrequestreview-893362658). Do you still have time to spend on this or should it rather be something we do in an another PR?
> 
> I can work as @chicco785 suggested in another PR.

ok!

**Review by chicco785 — APPROVED** [review]
